packages/financial-templates-lib/src/price-feed/LPPriceFeed.js

```diff
@@ -0,0 +1,129 @@
+   * @param {Function} getTime Returns the current time.
+   * @param {Function} [blockFinder] Optionally pass in a shared blockFinder instance (to share the cache).
+   * @param {Integer} [minTimeBetweenUpdates] Minimum amount of time that must pass before update will actually run
+   *                                        again.
+   * @param {Integer} [priceFeedDecimals] Precision that the caller wants precision to be reported in.
+   * @return None or throws an Error.
+   */
+  constructor({
+    logger,
+    web3,
+    erc20Abi,
+    poolAddress,
+    tokenAddress,
+    getTime,
+    blockFinder,
+    minTimeBetweenUpdates = 60,
+    priceFeedDecimals = 18
+  }) {
+    super();
+
+    // Assert required arguments.
+    assert(logger, "logger required");
+    assert(web3, "web3 required");
+    assert(erc20Abi, "erc20Abi required");
+    assert(poolAddress, "poolAddress required");
+    assert(tokenAddress, "tokenAddress required");
+    assert(getTime, "getTime required");
+
+    this.logger = logger;
+    this.web3 = web3;
+    this.toBN = web3.utils.toBN;
+
+    this.pool = new web3.eth.Contract(erc20Abi, poolAddress);
+    this.token = new web3.eth.Contract(erc20Abi, tokenAddress);
+    this.uuid = `LP-${poolAddress}-${tokenAddress}`;
+    this.getTime = getTime;
+    this.priceFeedDecimals = priceFeedDecimals;
+    this.minTimeBetweenUpdates = minTimeBetweenUpdates;
+    this.blockFinder = blockFinder || BlockFinder(web3.eth.getBlock);
+  }
+
+  getCurrentPrice() {
+    return this.price;
+  }
+
+  async getHistoricalPrice(time) {
+    const block = await this.blockFinder.getBlockForTimestamp(time);
+    return this._getPrice(block.number);
+  }
+
+  getLastUpdateTime() {
+    return this.lastUpdateTime;
+  }
+
+  getLookback() {
+    // Return infinity since this price feed can technically look back as far as needed.
+    return Infinity;
+  }
+
+  getPriceFeedDecimals() {
+    return this.priceFeedDecimals;
+  }
+
+  async update() {
+    const currentTime = await this.getTime();
+    if (this.lastUpdateTime === undefined || currentTime >= this.lastUpdateTime + this.minTimeBetweenUpdates) {
+      this.price = await this._getPrice();
+      this.lastUpdateTime = currentTime;
+    }
+  }
+
+  async _getDecimals() {
+    if (this.tokenDecimals === undefined) {
+      this.tokenDecimals = parseInt(await this.token.methods.decimals().call());
+    }
+
+    if (this.poolDecimals === undefined) {
+      this.poolDecimals = parseInt(await this.pool.methods.decimals().call());
+    }
+
+    return { poolDecimals: this.poolDecimals, tokenDecimals: this.tokenDecimals };
+  }
+
+  async _getPrice(blockNumber = "latest") {
+    const lpTotalSupply = this.toBN(await this.pool.methods.totalSupply().call(undefined, blockNumber));
+    const tokensInPool = this.toBN(
+      await this.token.methods.balanceOf(this.pool.options.address).call(undefined, blockNumber)
```

Comment:
How is it this has infinite lookback, is it only with archive node?

Reply:
Yeah, good point. For now, we assume the lookback is infinite because we don't know what type of node we're connected to.